---
title: "RBL3"
output: html_document
editor_options: 
  chunk_output_type: console
---
# Project Title
Drosha regulates mRNA splicing or decay
## Description
The project is to investigate Drosha-regulated genes and the mRNA splicing pattern in Drosha-KO.


I have wildtype cells and Drosha-KO, Dicer-KO, DGCR8-KO, Drosha-Dicer-DKO and DGCR8-Dicer-DKO cells. I would like to check which genes are regulated by Drosha, and whether Drosha will change the mRNA splicing pattern using Nanopore mRNA-seq .
Specifically, I would like to check whether Drosha regulates IGF2 by mRNA decay or alternative splicing. Thus, my question is: can I use Nanopore to detect the mRNA decay and alternative splicing by using a bioinformatics approach?

```{r}
library(dplyr)
library(ggplot2)
library(tidyverse)
library(ggpubr)

#references
#https://www.biorxiv.org/content/10.1101/672931v2.supplementary-material
```


```{r setup, include=FALSE}
data_dir = "/Volumes/projects/rbl3/tutorial/"
```

#abundance tables
```{r}
#plot transcript read lengths, read counts, exon counts
TranscriptFeatures <- function(df_in,title_in) {
  abund_df = read.csv(paste0(data_dir,df_in),sep="\t")
  
  p1 = ggplot(abund_df, aes(x=transcript_novelty, y=length, fill=transcript_novelty)) + 
    geom_violin(trim=FALSE)+
    geom_boxplot(width=0.1, fill="white")+
    labs(title="Plot of read length",
         x="Transcript type", y = "Read Length")+
    scale_fill_brewer(palette="RdBu") + theme_minimal()
  
  p2 = ggplot(abund_df, aes(x=transcript_novelty, y=n_exons, fill=transcript_novelty)) + 
    geom_violin(trim=FALSE)+
    geom_boxplot(width=0.1, fill="white")+
    labs(title="Plot of number of exons",
         x="Transcript Type", y = "Number of Exons")+
    scale_fill_brewer(palette="RdBu") + theme_minimal()

  transcript_nov = aggregate(SIRV_Rep1~transcript_novelty,abund_df,sum)
  transcript_nov$SIRV_Rep2 = aggregate(SIRV_Rep2~transcript_novelty,abund_df,sum)$SIRV_Rep2
  
  p3 = ggplot(data = transcript_nov %>% gather(Variable, SIRV_Rep1, -transcript_novelty), 
         aes(x = transcript_novelty, y = SIRV_Rep1, fill = Variable)) + 
      geom_bar(stat = 'identity', position = 'dodge') + 
    labs(y="Read Counts") + 
    ggtitle("Plot of read counts")

  p4 = ggparagraph(
    paste0("\n\n\n",title_in),
    color = NULL,
    size = 20,
    face = "bold",
    family = NULL,
    lineheight = NULL
  )
  
  #plot
  ggarrange(p1, p2, p3, p4,
            labels = c("A", "B", "C"),
            ncol = 2, nrow = 2)
} 


#plot unfiltered
TranscriptFeatures("example_talon_abundance.tsv","    Unfiltered Data")

#plot filtered; either:
#- transcripts are known OR
#- reads must have <50% A in the 20bp interval after alignmnet and 
#be detected at least 5 times in both replicates
#ISM - incomplete splice match; NIC - novel in catalog; NNC - novel not in catalog
TranscriptFeatures("example_talon_abundance_filtered.tsv","     Filtered Data")


```

```{r}
#reads messanger? full length mRNA? truncated mRNA?
```

